File: problems/01-arrays-and-hashing/003-valid-anagram/solution.go
/*
Problem: Valid Anagram
Difficulty: Easy
Pattern: Character Frequency Counting

Time Complexity Goal: O(n)
Space Complexity Goal: O(1)
*/

package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"strings"
)

/*
isAnagram checks if t is an anagram of s

TODO: Implement using map[rune]int for frequency counting
Hint: rune is used for Unicode characters in Go
*/
func isAnagram(s string, t string) bool {
	// TODO: Implement your solution here
	
	return false
}

type TestCase struct {
	ID          int    `json:"id"`
	Description string `json:"description"`
	Input       struct {
		S string `json:"s"`
		T string `json:"t"`
	} `json:"input"`
	Expected bool `json:"expected"`
}

type TestData struct {
	Problem   string     `json:"problem"`
	TestCases []TestCase `json:"testCases"`
}

func runTests() {
	data, err := ioutil.ReadFile("test-data.json")
	if err != nil {
		fmt.Printf("Error reading test file: %v\n", err)
		return
	}

	var testData TestData
	json.Unmarshal(data, &testData)

	fmt.Println(strings.Repeat("=", 60))
	fmt.Println("Running Valid Anagram Tests - Go")
	fmt.Println(strings.Repeat("=", 60))

	passed, failed := 0, 0

	for _, test := range testData.TestCases {
		s := test.Input.S
		t := test.Input.T
		expected := test.Expected

		result := isAnagram(s, t)

		if result == expected {
			fmt.Printf("✅ Test %d: PASSED\n", test.ID)
			passed++
		} else {
			fmt.Printf("❌ Test %d: FAILED\n", test.ID)
			fmt.Printf("   Input: s=\"%s\", t=\"%s\"\n", s, t)
			fmt.Printf("   Expected: %v, Got: %v\n", expected, result)
			failed++
		}
		fmt.Println()
	}

	fmt.Println(strings.Repeat("=", 60))
	fmt.Printf("Results: %d passed, %d failed\n", passed, failed)
	fmt.Println(strings.Repeat("=", 60))

	if passed == len(testData.TestCases) {
		fmt.Println("🎉 All tests passed!")
	}
}

func main() {
	fmt.Println("Quick Test: s=\"anagram\", t=\"nagaram\" -> Expected: true")
	fmt.Printf("Your output: %v\n", isAnagram("anagram", "nagaram"))
	fmt.Println()

	// runTests()
}

File: problems/01-arrays-and-hashing/003-valid-anagram/solution.cs
/*
 * Problem: Valid Anagram
 * Difficulty: Easy
 * Pattern: Character Frequency Counting
 * 
 * Time Complexity Goal: O(n)
 * Space Complexity Goal: O(1)
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Linq;

public class ValidAnagramSolution
{
        /// <summary>
        /// Check if t is an anagram of s
        /// </summary>
        public static bool IsAnagram(string s, string t)
        {
            // TODO: Implement using Dictionary<char, int>
            // Hint: LINQ can help with sorting approach
            // Alternative: Count frequencies manually
            
            return false;
        }

        public class TestCase
        {
            public int id { get; set; }
            public string description { get; set; }
            public TestInput input { get; set; }
            public bool expected { get; set; }
        }

        public class TestInput
        {
            public string s { get; set; }
            public string t { get; set; }
        }

        public class TestData
        {
            public string problem { get; set; }
            public List<TestCase> testCases { get; set; }
        }

        public static void RunTests()
        {
            string jsonString = File.ReadAllText("test-data.json");
            var data = JsonSerializer.Deserialize<TestData>(jsonString, new JsonSerializerOptions
            {
                PropertyNameCaseInsensitive = true
            });

            Console.WriteLine(new string('=', 60));
            Console.WriteLine("Running Valid Anagram Tests - C#");
            Console.WriteLine(new string('=', 60));

            int passed = 0, failed = 0;

            foreach (var test in data.testCases)
            {
                var s = test.input.s;
                var t = test.input.t;
                var expected = test.expected;

                try
                {
                    var result = IsAnagram(s, t);

                    if (result == expected)
                    {
                        Console.WriteLine($"✅ Test {test.id}: PASSED");
                        passed++;
                    }
                    else
                    {
                        Console.WriteLine($"❌ Test {test.id}: FAILED");
                        Console.WriteLine($"   Input: s=\"{s}\", t=\"{t}\"");
                        Console.WriteLine($"   Expected: {expected}, Got: {result}");
                        failed++;
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"❌ Test {test.id}: ERROR - {ex.Message}");
                    failed++;
                }

                Console.WriteLine();
            }

            Console.WriteLine(new string('=', 60));
            Console.WriteLine($"Results: {passed} passed, {failed} failed");
            Console.WriteLine(new string('=', 60));

            if (passed == data.testCases.Count)
            {
                Console.WriteLine("🎉 All tests passed!");
            }
        }

}

// Quick manual test (global code for dotnet script)
Console.WriteLine("Quick Test: s=\"anagram\", t=\"nagaram\" -> Expected: True");
Console.WriteLine($"Your output: {ValidAnagramSolution.IsAnagram(\"anagram\", \"nagaram\")}");
Console.WriteLine();

// Uncomment to run all tests
// ValidAnagramSolution.RunTests();

